src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs

```diff
@@ -459,31 +463,99 @@ private void StartTelemetryIfFunctionInvocation(IDictionary<string, object> stat
                     !string.IsNullOrEmpty(functionInvocationId) &&
                     eventName == LogConstants.FunctionStartEvent)
                 {
-                    RequestTelemetry request = new RequestTelemetry
+                    IOperationHolder<RequestTelemetry> operation;
+
+                    // link represents context from the upstream service that is not necessarily immediate parent
+                    // it is used by EventHubs to represent context in the message.
+                    // if there is just one link, we'll use it as a parent as an optimization.
+                    // if there is more than one, we'll populate them as custom properties
+                    IEnumerable<Activity> links = allScopes.GetValueOrDefault<IEnumerable<Activity>>("Links");
+                    var activities = links as Activity[] ?? links?.ToArray();
+
+                    if (activities != null && activities.Length == 1)
+                    {
+                        operation = _telemetryClient.StartOperation<RequestTelemetry>(activities[0]);
+                        operation.Telemetry.Name = functionName;
+                    }
+                    else
                     {
-                        Name = functionName
-                    };
+                        RequestTelemetry request = new RequestTelemetry
+                        {
+                            Name = functionName
+                        };
 
-                    // We'll need to store this operation context so we can stop it when the function completes
-                    IOperationHolder<RequestTelemetry> operation = _telemetryClient.StartOperation(request);
+                        operation = _telemetryClient.StartOperation(request);
 
+                        // if there is more than one link (batch dispatch in EventHub trigger)
+                        // we'll populate link information on the request telemetry, but don't touch parent for this request
+                        PopulateLinks(activities, request);
+                    }
+
+                    // We'll need to store this operation context so we can stop it when the function completes
                     stateValues[OperationContext] = operation;
                 }
             }
         }
 
+        private void PopulateLinks(Activity[] activities, RequestTelemetry request)
+        {
+            if (activities == null)
+            {
+                return;
+            }
+
+            if (activities.Any(l => l.Recorded))
+            {
+                request.ProactiveSamplingDecision = SamplingDecision.SampledIn;
+            }
```

Comment:
What does Recorded mean?
Also, you can force something to be included even if Sampling is on?

Reply:
Recoded flag on activity means that parent was recorded (sampled in).

In event hubs scenarios, we may get 10 different unrelated messages with different contexts. We know which of there were sampled in by upstream service.

If any of the messages is sampled in, we mark request telemtery (for trigger) to be proactively sampled in.

It is processed once again when request is being tracked (function completes) and sampling algorithm will give more priority to telemtery that was proactively sampled in (while still keeping the rate under configured one).

It actually reminded me that we need to set Activity.Recorded to true on the Activity we create for trigger so everything under it's scope will get higher priority  - I'll add it and more tests for it